### PR TITLE
adds graceful failure mode; all public functions have a try/catch (FF…

### PR DESCRIPTION
**Pull request by leoromanovsky**
…-968)

---
labels: mergeable
---
[//]: #  (Link to the issue corresponding to this chunk of work)
Fixes: #__issue__

## Motivation and Context
[//]: #  (Why is this change required? What problem does it solve?)

Exceptions within the Eppo SDK should not bubble up to the customers' applications.

## Description
[//]: # (Describe your changes in detail)

By default the graceful mode is enabled; all exceptions are captured in the top level public methods and logged.

For development the graceful mode can be disabled:

```
client.setIsGracefulFailureMode(false)
```

## How has this been tested?
[//]: # (Please describe in detail how you tested your changes)

It was easy to test the logger because we can pass a mock - it wasn't obvious to me how to trigger an exception to be able to test the graceful mode. We'd need to stub `getAssignmentVariation` but this seems unnecessary.

Suggestions welcome.

## deployment plan

Bump version number and import into upstream.

[//]: # (OPTIONAL)
[//]: # (Add one or multiple labels: enhancement, refactoring, bugfix)

### CODE REVIEW

File: src/client/eppo-client.ts [hunk 1]
```diff
@@ -113,15 +122,23 @@ export default class EppoClient implements IEppoClient {
...
-        subjectKey,
-        flagKey,
-        subjectAttributes,
-        assignmentHooks,
-        ValueType.StringType,
-      ).stringValue ?? null
-    );
+    try {
+      return (
+        this.getAssignmentVariation(
```
Comment (aarsilv) [inline src/client/eppo-client.ts:127]: Since all these functions do is immediately delegate to `getAssignmentVariation()`, I'm thinking it would be cleaner to just have the `try`-`catch` in that method and forgo it in these user-facing syntactic sugar wrappers.

File: src/client/eppo-client.ts [hunk 2]
```diff
@@ -100,10 +101,18 @@ export default class EppoClient implements IEppoClient {
...
-        .stringValue ?? null
-    );
+    try {
+      return (
+        this.getAssignmentVariation(subjectKey, flagKey, subjectAttributes, assignmentHooks)
+          .stringValue ?? null
+      );
+    } catch (error) {
+      if (this.isGracefulFailureMode) {
+        console.error(`[Eppo SDK] Error getting assignment: ${error.message}`);
```
Comment (aarsilv) [inline src/client/eppo-client.ts:111]: Consider just logging the error so we get the stack trace as well.
```ts
console.error('[Eppo SDK] Error getting assignment', error.message);
```

**Review by aarsilv — COMMENTED** [review]
I'd include unit tests that mock out bad calls for both failures modes to verify exception is thrown / null returned.

File: src/client/eppo-client.ts [hunk 3]
```diff
@@ -288,6 +337,10 @@ export default class EppoClient implements IEppoClient {
     this.flushQueuedEvents(); // log any events that may have been queued while initializing
   }
 
+  public setIsGracefulFailureMode(gracefulFailureMode: boolean) {
+    this.isGracefulFailureMode = gracefulFailureMode;
+  }
```
[the remark this answers is not in the document]
Comment (leoromanovsky) [reply src/client/eppo-client.ts:342]: yea thank you - good idea!

File: src/client/eppo-client.spec.ts [hunk 4]
```diff
@@ -130,6 +130,67 @@ describe('EppoClient E2E test', () => {
     },
   };
 
+  describe('error encountered', () => {
```
Comment (aarsilv) [inline src/client/eppo-client.spec.ts:133]: 🙌

**Review by aarsilv — APPROVED** [review]
Thanks for pairing and iterating! 🎉
